iOS: Keys and EphIDs generation

loicgardiol opened the pull request:
Still independent code. I haven't removed the existing code and no parts of the app relies on this new code.

loicgardiol committed:
- Import and adapt SK and EphID generation from DP3T (EPFL design)
- SecurePersistence convenience class for storing objects in keychain

loicgardiol committed:
Remove old file header

loicgardiol committed:
Add documentation

ubamrein commented on Sources/STARSDK/Cryptography/CryptoParams.swift:
```diff
@@ -0,0 +1,12 @@
+struct GlobalParameters {
+    static let epoch0TimeInterval: TimeInterval = 1585475061.0
+    static let epochDuration: TimeInterval = 15.0 * 60.0
+    static let nbStoredEpochs: UInt32 = 21
```
Maybe we should rename that. I don't see how this value is related to epochs (we have 96 epochs a day right?). Something like storedDailySecretKeys or similiar.

loicgardiol replied:
Coming from the prototype code, this was for code simplicity in the epoch rotation code.
But sure, we can put a number of days and compute the resulting number of epochs in the code rotation part :-)

ubamrein replied:
We don't have to compute anything right? I think the name is just wrong :) 
This value is used to determine, if we should drop SecretKeys (we have one a day right?). So it has nothing to do with epochs (we have 96 epochs a day each of which lasts for 15 minutes), or am I confusing something?
Hence, I would suggest to use nbStoredDailySecrets instead.

loicgardiol replied:
You are right. Sorry, I got stuck on the previous design, where there was 1 Secret Key per epoch. Then the code for has to changed, it still implements the previous design.

ubamrein commented on Sources/STARSDK/Cryptography/IdentityManager.swift:
```diff
@@ -0,0 +1,163 @@
+private class Utils {
+    /// Generates a random key using CommonCrypto.
+    /// Assumes nbBits is a multiple of 8.
+    static func generateRandomKey(nbBits: Int) -> Data {
```
Is there a particular reason why we want bits and not bytes? It seems we can anyways only generate a multiple of 8bits. if we really want bits, we should use floor for the number of bytes generated. Otherwise we might get OutOfBounds (e.g. we want 9bits but there are only 8bits generated)

loicgardiol replied:
Went with bits because keys are typically expressed in bits, I'm totally fine to move to bytes for the reason you mention.

stmitt commented on Sources/STARSDK/Cryptography/IdentityManager.swift:
```diff
@@ -0,0 +1,163 @@
+    }
+}
+
+struct SK: Codable, CustomStringConvertible {
```
can we rename SK to SecretKey for better code readability

loicgardiol replied:
Sure. Use `SK` because this is the given name in the EPFL spec.

loicgardiol committed:
iOS:
- Rename SK to SecretKey
- Change from bits to bytes for key lengths
- Add possibility to stop and invalidate EpochTracker

ubamrein reviewed (changes requested):
Currently this PR implements the wrong protocol. 

So for clarification:
- We have `1` SecretKey per day, where each day has `96` Epochs (a new Epoch every `15` minutes).
- Every successive day uses the SecretKey of the previous day to generate a new SecretKey (SHA256).
- Every Epoch has an identifier derived from the daily SecretKey (have we decided, if we use a HTOP approach like in our original one, or an AES Stream-Cipher as in your white paper?)

loicgardiol commented:
Current implementation implements the old design (when there was one secret key per epoch). I will re-write that tonight.
